### ISSUE TITLE
Help with jax optimization routine for image warping 

### ISSUE THREAD

**Issue by anorak94** [google/jax#5248]
Hey i am trying to use jax to compute warping fields for two images. I have a circle and a square and I would like to warp the circle to the square. Right now i am using a very simple least squares approach to the problem.

```python
def identity_mapping(shape):
    """
    initializing the parameters 
    """
    x1 = [float(i) for i in range(shape[0])]
    y1 = [float(i)for i in range(shape[1])]
    return jax.numpy.meshgrid(x1, y1, indexing = "xy")
    

def similarity(phi):
    """
    warp the original image and compute the least squares error
    """
    warped = jax.scipy.ndimage.map_coordinates(circle, phi, order=1)
    sim = jnp.sum(jnp.square((warped - square)))
    return sim

@jit
def update(step, opt_state):
    """
    updating the parameters values
    """
    value, grads = value_and_grad(similarity)(get_params(opt_state))
    opt_state = opt_update(step, grads, opt_state)
    return value, opt_state, grads


step_size = 10
num_epochs = 40


opt_init, opt_update, get_params = optimizers.adam(step_size)
params = identity_mapping(circle.shape)
opt_state = opt_init(params)

loss = []
for epoch in tqdm(range(num_epochs)):
    #print (epoch)
    value, opt_state, grad = update(epoch, opt_state)
    loss.append(value)  


```

But the problem is that my gradient updates are going to zero in most cases as the loss function evaluates to a constant. usually there are smoothness constraints on the deformation mapping as well but i am skipping that for now. I was wondering if the way i am doing the gradient updates are correct or not. I have posted the images below

![circle (1)](https://user-images.githubusercontent.com/45297533/102999852-1ba9fc00-452a-11eb-959b-4687a46fc645.png)
![square (1)](https://user-images.githubusercontent.com/45297533/102999855-1cdb2900-452a-11eb-9381-458819eca648.png)

**Comment by hawkinsp** [comment]
At a first glance, the optimizer update looks fine. If I had to guess, I suspect taking the gradient of `map_coordinates` doesn't work as well as you would like, since it will cast the coordinates to integers for indexing purposes. Are you trying to implement a well-known algorithm here? If so, how is the gradient of the coordinate mapping usually computed?

**Comment by anorak94** [comment]
hey i managed to make it work. the code is fine. i had to do execute it in a multi-resolution scheme it seems to work then.
